NExtends/Expressions/ExpressionEqualityComparer.cs

```diff
@@ -0,0 +1,239 @@
+		{
+			if (ReferenceEquals(x, y)) return true;
+			if (x == null || y == null) return false;
+
+			var valueX = TryCalculateConstant(x);
+			var valueY = TryCalculateConstant(y);
+
+			if (valueX.IsDefined && valueY.IsDefined)
+				return ValuesEqual(valueX.Value, valueY.Value);
+
+			if (x.NodeType != y.NodeType
+				|| x.Type != y.Type)
+			{
+				if (IsAnonymousType(x.Type) && IsAnonymousType(y.Type))
+					throw new NotImplementedException("Comparison of Anonymous Types is not supported");
+				return false;
+			}
+
+			if (x is LambdaExpression)
+			{
+				var lx = (LambdaExpression)x;
+				var ly = (LambdaExpression)y;
+				var paramsX = lx.Parameters;
+				var paramsY = ly.Parameters;
+				return CollectionsEqual(paramsX, paramsY, lx, ly) && ExpressionsEqual(lx.Body, ly.Body, lx, ly);
+			}
+			if (x is MemberExpression)
+			{
+				var mex = (MemberExpression)x;
+				var mey = (MemberExpression)y;
+				return Equals(mex.Member, mey.Member) && ExpressionsEqual(mex.Expression, mey.Expression, rootX, rootY);
+			}
+			if (x is BinaryExpression)
+			{
+				var bx = (BinaryExpression)x;
+				var by = (BinaryExpression)y;
+				return bx.Method == @by.Method && ExpressionsEqual(bx.Left, @by.Left, rootX, rootY) &&
+					   ExpressionsEqual(bx.Right, @by.Right, rootX, rootY);
+			}
+			if (x is UnaryExpression)
+			{
+				var ux = (UnaryExpression)x;
+				var uy = (UnaryExpression)y;
+				return ux.Method == uy.Method && ExpressionsEqual(ux.Operand, uy.Operand, rootX, rootY);
+			}
+			if (x is ParameterExpression)
+			{
+				var px = (ParameterExpression)x;
+				var py = (ParameterExpression)y;
+				return rootX.Parameters.IndexOf(px) == rootY.Parameters.IndexOf(py);
+			}
+			if (x is MethodCallExpression)
+			{
+				var cx = (MethodCallExpression)x;
+				var cy = (MethodCallExpression)y;
+				return cx.Method == cy.Method
+					   && ExpressionsEqual(cx.Object, cy.Object, rootX, rootY)
+					   && CollectionsEqual(cx.Arguments, cy.Arguments, rootX, rootY);
+			}
+			if (x is MemberInitExpression)
+			{
+				var mix = (MemberInitExpression)x;
+				var miy = (MemberInitExpression)y;
+				return ExpressionsEqual(mix.NewExpression, miy.NewExpression, rootX, rootY)
+					   && MemberInitsEqual(mix.Bindings, miy.Bindings, rootX, rootY);
+			}
+			if (x is NewArrayExpression)
+			{
+				var nx = (NewArrayExpression)x;
+				var ny = (NewArrayExpression)y;
+				return CollectionsEqual(nx.Expressions, ny.Expressions, rootX, rootY);
+			}
+			if (x is NewExpression)
+			{
+				var nx = (NewExpression)x;
+				var ny = (NewExpression)y;
+				return
+					Equals(nx.Constructor, ny.Constructor)
+					&& CollectionsEqual(nx.Arguments, ny.Arguments, rootX, rootY)
+					&& (nx.Members == null && ny.Members == null
+						|| nx.Members != null && ny.Members != null && CollectionsEqual(nx.Members, ny.Members));
+			}
+			if (x is ConditionalExpression)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this type-check-and-cast sequence with an 'as' and a null check. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3247)